### ISSUE TITLE
MIDI learn does not work properly in Helm

### ISSUE THREAD

**Issue by riban-bw** [zynthian/zynthian-issue-tracking#266]
**Describe the bug**
MIDI learn to control parameters from MIDI continuous controllers always sets Helm parameters to same value.

**To Reproduce**
Steps to reproduce the behaviour:
1. Add Helm layer
2. Select a control to MIDI learn
3. Adjust CC 1 to set it as controller

**Expected behaviour**
Helm parameter changes between its minimum and maximum when MIDI CC values 0 to 127 are received.

**Actual behaviour**
Helm parameter is set to a static value that never changes.

**Setup:**
HARDWARE
Raspberry Pi 4 Model B Rev 1.2
Soundcard: Behringer UMC1820
Display: WaveShare 7 HDMI+USB
Wiring: I2C_HWC
GPIO Expander: Not detected

SYSTEM
Raspbian GNU/Linux 10 (buster)
Build Date: 2020-04-09
Memory: 6% (270M/3906M)
SD Card: 39% (11G/28G)
Temperature: 54.0ºC

SOFTWARE
zyncoder: master (517e46a)
zynthian-ui: master (8ba25fb)
zynthian-sys: master (4ea75d9)
zynthian-data: master (20e7da0)
zynthian-webconf: master (27672ca)

**Comment by riban-bw** [comment]
Update: This seems to only affect CC 1. Other CC work fine. I wonder if there is some limitation in Helm or Zynthian?

**Comment by jofemodo** [comment]
I just tried and it works OK for me. I'm using an "AKAI MPK mini" as keyboard/controller.
I can map any parameter to any CC, including CC1. 

BTW, have you checked with the MIDI log?

Regards,

**Comment by riban-bw** [comment]
Something rather odd going on here... Connecting the MIDI output of my VZ1 via MIDISPORT2x2 USB interface to GMIDImonitor (using JACK) shows my controller is sending CC1 values 0..127 on channel 1. Looking at Zynthian MIDI log monitoring ZynMidiRouter Main shows CC74 values 0..127 on channel 1. Looking at Zynthian MIDI log monitoring ZynMidiRouter MidiSport 2x2 MIDI 1 shows CC1 values 0..127 on channel 1. So the CC is being transformed within Zynthian somewhere. This happens for CC1->CC74, CC31->CC71. CC 12..30 all pass through correctly. (I can't test other values with this controller.) I have removed all MIDI filter rules and rebooted but same effect. So that is odd but the original report persists with different / additional info: Helm configured to MIDI learn from CC74 or CC71 sets a static value. Other engines do work with those controller numbers.

The issue with weird mapping of CC numbers may need to be reported separately.

**Comment by riban-bw** [comment]
Removing Layer 1 resolves this bad behaviour. Snapshot attached.
[last_state.zip](https://github.com/zynthian/zynthian-issue-tracking/files/5224914/last_state.zip)

**Comment by riban-bw** [comment]
Further info: It looks like assigning the same CC on the same MIDI channel to parameters on multiple engines causes this behaviour.

**Comment by jofemodo** [comment]
I just tried to create 2 layers and assign CC1 to several parameters on both of them. It worked like a charm!

Perhaps you have some old MIDI filter ruleset embeded on your snapshot. This is an old problem i have to solve.

The problem is that MIDI filter rules are saved with snapshot, but when you load the snapshot, these rules are hidden. Webconf display current MIDI profile's filter rules and no the snapshot's rules, so you can forget that the snapshot have these rules. We should revise the "MIDI profile" concept and perhaps, bury it.

**Comment by riban-bw** [comment]
I can repeat this for many different synth engines, e.g. Fluidsynth + AMSynth. I have to be in multitimbral mode, i.e. disable "Single Channel" mode. It looks like an attempt to assign a controller that is already assigned to a different layer fails (and the default controller remains shown) but the value of the parameter changes as you do this and subsequent attempts to use the control show in MIDI log with CC number that is the default for the parameter you tried to set, e.g.

- Add Fluidsynth as layer 1
- Add AMSynth as layer 2
- Assign CC1 to control layer 1 (fluidsynth) parameter "Pan" - success
- Assign CC1 to control layer 2 (AMSynth) parameter "Amp Sustain" - fail
AMSynth Amp Sustain value changes but the assigned controller returns to its default of CC10. 
- Adjust CC1 and observe MIDI log that CC10 is being logged

**Comment by jofemodo** [comment]
OK! I catched it. It's a limitation of the current implementation of MIDI learning on some engines. If i'm right, it should happens when using some of these engines:

+ FluidSynth
+ LinuxSampler
+ setBfree
+ Pianoteq

in combination with others.

Please, if you could confirm, i'm thinking about how to solve it ...

Thanks!

**Comment by riban-bw** [comment]
I also saw it with zynaddsubfx and amsynth.

**Comment by jofemodo** [comment]
Yes, you are right. The pre-asigned controllers from ZynAddSubFX also have "the problem".
